metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,330 @@
+import os
+import sys
+import signal
+import psutil
+import shutil
+import asyncio
+import tempfile
+from typing import List, Dict, Optional, Callable
+
+
+def kill_process_and_descendants(pid, termination_timeout):
+    def on_terminate(proc):
+        print("process %s terminated" % proc)
+
+    try:
+        parent = psutil.Process(pid)
+        children = parent.children(recursive=True)
+        for p in children:
+            p.terminate()
+        _, alive = psutil.wait_procs(
+            children, timeout=termination_timeout, callback=on_terminate
+        )
+        for p in alive:
+            p.kill()
+    except psutil.NoSuchProcess:
+        pass
+
+
+class LogReadTimeoutError(Exception):
+    """Exception raised when reading logs times out."""
+
+    pass
+
+
+class SubprocessManager(object):
```

Comment:
If we are exposing this to user, please add docstrings. I am not sure we are though in which case we can add to make clean but not as mandatory.

Reply:
they are already present, but we can improve them in a different PR